research/minigo/dual_net.py

```diff
@@ -0,0 +1,209 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Contains utility and supporting functions for DualNet.
+
+This module provides the model interface, including functions for DualNet model
+bootstrap, training, validation, loading and exporting.
+"""
+
+# pylint: disable=g-bad-import-order
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+
+import tensorflow as tf
+
+import dualnet_model
+import features
+import go
+import model_utils
+import preprocessing
+import symmetries
+
+
+class DualNetwork(object):
+  """The DualNetwork class for the complete model with graph and weights.
+
+  This class can restore the model from saved files, and provide inference for
+  given examples.
+  """
+
+  def __init__(self, save_file, params):
+    """Initialize the dual network from saved model/checkpoints.
+
+    Args:
+      save_file: Path where model parameters were previously saved.
+      params: Hyper parameters for the dual network
+    """
+    self.save_file = save_file
+    self.hparams = params
+    self.inference_input = None
+    self.inference_output = None
+    config = tf.ConfigProto()
+    config.gpu_options.allow_growth = True
+    self.sess = tf.Session(graph=tf.Graph(), config=config)
+    self.initialize_graph()
+
+  def initialize_graph(self):
+    with self.sess.graph.as_default():
+      input_features, labels = get_inference_input()
+      estimator_spec = dualnet_model.model_fn(
+          input_features, labels, tf.estimator.ModeKeys.PREDICT, self.hparams)
+      self.inference_input = input_features
+      self.inference_output = estimator_spec.predictions
+      if self.save_file is not None:
+        self.initialize_weights(self.save_file)
+      else:
+        self.sess.run(tf.global_variables_initializer())
+
+  def initialize_weights(self, save_file):
+    """Initialize the weights from the given save_file.
+
+    Assumes that the graph has been constructed, and the save_file contains
+    weights that match the graph. Used to set the weights to a different version
+    of the player without redefining the entire graph.
+
+    Args:
+      save_file: Path where model parameters were previously saved.
+    """
+
+    tf.train.Saver().restore(self.sess, save_file)
+
+  def run(self, position, use_random_symmetry=True):
+    """Compute the policy and value output for a given position."""
+    probs, values = self.run_many(
+        [position], use_random_symmetry=use_random_symmetry)
+    return probs[0], values[0]
+
+  def run_many(self, positions, use_random_symmetry=True):
+    processed = list(map(features.extract_features, positions))
```

Comment:
Ditto the above on docstrings. Also, comment on what the expected values are here-- what features are extracted from the positions, what is the expected format, how are they used?